# Przedstawienie 5 najniebezpieczniejszych czynników wypadków pod względem śmiertelności w NY

## Wczytywanie danych z pliku CSV

In [1]:
import pandas as pd
import numpy as np

import requests

nypd_df = pd.read_csv("/home/student/nypd-motor-vehicle-collisions.csv")
nypd_df

/tmp/ipykernel_87204/3460021377.py:6: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  nypd_df = pd.read_csv("/home/student/nypd-motor-vehicle-collisions.csv")


ACCIDENT DATE ACCIDENT TIME    BOROUGH ZIP CODE   LATITUDE   
0        2019-08-05T00:00:00.000         16:30     QUEENS  11434.0  40.676052  \
1        2019-08-27T00:00:00.000         16:02   BROOKLYN  11225.0  40.657780   
2        2019-08-15T00:00:00.000         17:57  MANHATTAN  10002.0  40.718143   
3        2019-08-30T00:00:00.000         21:53      BRONX  10460.0  40.840534   
4        2019-08-06T00:00:00.000          9:45  MANHATTAN  10016.0  40.745440   
...                          ...           ...        ...      ...        ...   
1612173  2012-07-21T00:00:00.000         22:46   BROOKLYN  11226.0  40.651741   
1612174  2012-07-09T00:00:00.000         16:45        NaN      NaN        NaN   
1612175  2012-07-09T00:00:00.000         14:35        NaN      NaN  40.791492   
1612176  2012-07-18T00:00:00.000          9:20  MANHATTAN  10024.0  40.779479   
1612177  2012-07-12T00:00:00.000          3:10     QUEENS  11361.0  40.762501   

         LONGITUDE                                           LOCATION   
0       -73.790184  {'type': 'Point', 'coordinates': [-73.790184, ...  \
1       -73.951096  {'type': 'Point', 'coordinates': [-73.951096, ...   
2       -73.993835  {'type': 'Point', 'coordinates': [-73.993835, ...   
3       -73.866610  {'type': 'Point', 'coordinates': [-73.86661, 4...   
4       -73.975400  {'type': 'Point', 'coordinates': [-73.9754, 40...   
...            ...                                                ...   
1612173 -73.965241  {'type': 'Point', 'coordinates': [-73.9652409,...   
1612174        NaN                                                NaN   
1612175 -73.935763  {'type': 'Point', 'coordinates': [-73.935763, ...   
1612176 -73.973573  {'type': 'Point', 'coordinates': [-73.9735729,...   
1612177 -73.782522  {'type': 'Point', 'coordinates': [-73.7825216,...   

                           ON STREET NAME                 CROSS STREET NAME   
0                                     NaN                               NaN  \
1                                     NaN                               NaN   
2        CHRYSTIE STREET                                       GRAND STREET   
3                                     NaN                               NaN   
4        EAST 35 STREET                                            2 AVENUE   
...                                   ...                               ...   
1612173  PARADE PLACE                      CROOKE AVENUE                      
1612174  EDWARD L GRANT HIGHWAY            WEST 169 STREET                    
1612175                               NaN                               NaN   
1612176  CENTRAL PARK WEST                 WEST 77 STREET                     
1612177  39 AVENUE                         203 STREET                         

                                  OFF STREET NAME  ...   
0        150-08    123 AVENUE                      ...  \
1        288       HAWTHORNE STREET                ...   
2                                             NaN  ...   
3        1837      EAST TREMONT AVENUE             ...   
4                                             NaN  ...   
...                                           ...  ...   
1612173                                       NaN  ...   
1612174                                       NaN  ...   
1612175                                       NaN  ...   
1612176                                       NaN  ...   
1612177                                       NaN  ...   

          CONTRIBUTING FACTOR VEHICLE 2  CONTRIBUTING FACTOR VEHICLE 3   
0                           Unspecified                            NaN  \
1                           Unspecified                            NaN   
2                                   NaN                            NaN   
3                           Unspecified                            NaN   
4        Driver Inattention/Distraction                            NaN   
...                                 ...                            ...   
16121

##### Wczytanie tabeli ze wszytskimi danymi do analizy czynników wypadków w NY

## Wyznaczenie kolumn z niebezpiecznymi czynnikami wypadków, osób poszkodowanych oraz ich śmiertelnością

In [2]:
Selected_data_columns = nypd_df[['NUMBER OF PERSONS KILLED','CONTRIBUTING FACTOR VEHICLE 1','CONTRIBUTING FACTOR VEHICLE 2','CONTRIBUTING FACTOR VEHICLE 3','CONTRIBUTING FACTOR VEHICLE 4','CONTRIBUTING FACTOR VEHICLE 5']]
Selected_data_columns

NUMBER OF PERSONS KILLED   CONTRIBUTING FACTOR VEHICLE 1   
0                             0.0             Passing Too Closely  \
1                             0.0             Passing Too Closely   
2                             0.0  Driver Inattention/Distraction   
3                             0.0                     Unspecified   
4                             0.0  Driver Inattention/Distraction   
...                           ...                             ...   
1612173                       0.0     Traffic Control Disregarded   
1612174                       0.0                 Fatigued/Drowsy   
1612175                       0.0                 Fatigued/Drowsy   
1612176                       0.0  Driver Inattention/Distraction   
1612177                       0.0                     Unspecified   

          CONTRIBUTING FACTOR VEHICLE 2 CONTRIBUTING FACTOR VEHICLE 3   
0                           Unspecified                           NaN  \
1                           Unspecified                           NaN   
2                                   NaN                           NaN   
3                           Unspecified                           NaN   
4        Driver Inattention/Distraction                           NaN   
...                                 ...                           ...   
1612173                     Unspecified                           NaN   
1612174                 Fatigued/Drowsy                   Unspecified   
1612175                     Unspecified                           NaN   
1612176                     Unspecified                   Unspecified   
1612177                             NaN                           NaN   

        CONTRIBUTING FACTOR VEHICLE 4 CONTRIBUTING FACTOR VEHICLE 5  
0                                 NaN                           NaN  
1                                 NaN                           NaN  
2                                 NaN                           NaN  
3                                 NaN                           NaN  
4                                 NaN                           NaN  
...                               ...                           ...  
1612173                           NaN                           NaN  
1612174                           NaN                           NaN  
1612175                           NaN                           NaN  
1612176                           NaN                           NaN  
1612177                           NaN                           NaN  

[1612178 rows x 6 columns]

##### Kolumna z danymi do określenia 5 czynników wypadków oraz ich śmiertelność w NY

## Usunięcie wierszy, które w kolumnach osób zabitych mają nieokreśloną wartość (NaN)

In [3]:
nypd_factors = Selected_data_columns.dropna(subset=['NUMBER OF PERSONS KILLED'])
nypd_factors

NUMBER OF PERSONS KILLED   CONTRIBUTING FACTOR VEHICLE 1   
0                             0.0             Passing Too Closely  \
1                             0.0             Passing Too Closely   
2                             0.0  Driver Inattention/Distraction   
3                             0.0                     Unspecified   
4                             0.0  Driver Inattention/Distraction   
...                           ...                             ...   
1612173                       0.0     Traffic Control Disregarded   
1612174                       0.0                 Fatigued/Drowsy   
1612175                       0.0                 Fatigued/Drowsy   
1612176                       0.0  Driver Inattention/Distraction   
1612177                       0.0                     Unspecified   

          CONTRIBUTING FACTOR VEHICLE 2 CONTRIBUTING FACTOR VEHICLE 3   
0                           Unspecified                           NaN  \
1                           Unspecified                           NaN   
2                                   NaN                           NaN   
3                           Unspecified                           NaN   
4        Driver Inattention/Distraction                           NaN   
...                                 ...                           ...   
1612173                     Unspecified                           NaN   
1612174                 Fatigued/Drowsy                   Unspecified   
1612175                     Unspecified                           NaN   
1612176                     Unspecified                   Unspecified   
1612177                             NaN                           NaN   

        CONTRIBUTING FACTOR VEHICLE 4 CONTRIBUTING FACTOR VEHICLE 5  
0                                 NaN                           NaN  
1                                 NaN                           NaN  
2                                 NaN                           NaN  
3                                 NaN                           NaN  
4                                 NaN                           NaN  
...                               ...                           ...  
1612173                           NaN                           NaN  
1612174                           NaN                           NaN  
1612175                           NaN                           NaN  
1612176                           NaN                           NaN  
1612177                           NaN                           NaN  

[1612145 rows x 6 columns]

##### Oczyszczona tabela z niepotrzebnych(nieokreślonych) wartości

## Dodanie kolumny z czynnikiem wypadków

In [4]:
nypd_factors['CONTRIBUTING FACTORS'] = nypd_factors[['CONTRIBUTING FACTOR VEHICLE 1', 'CONTRIBUTING FACTOR VEHICLE 2', 'CONTRIBUTING FACTOR VEHICLE 3', 'CONTRIBUTING FACTOR VEHICLE 4', 'CONTRIBUTING FACTOR VEHICLE 5']].values.tolist()
nypd_factors

/tmp/ipykernel_87204/2732455402.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  nypd_factors['CONTRIBUTING FACTORS'] = nypd_factors[['CONTRIBUTING FACTOR VEHICLE 1', 'CONTRIBUTING FACTOR VEHICLE 2', 'CONTRIBUTING FACTOR VEHICLE 3', 'CONTRIBUTING FACTOR VEHICLE 4', 'CONTRIBUTING FACTOR VEHICLE 5']].values.tolist()


NUMBER OF PERSONS KILLED   CONTRIBUTING FACTOR VEHICLE 1   
0                             0.0             Passing Too Closely  \
1                             0.0             Passing Too Closely   
2                             0.0  Driver Inattention/Distraction   
3                             0.0                     Unspecified   
4                             0.0  Driver Inattention/Distraction   
...                           ...                             ...   
1612173                       0.0     Traffic Control Disregarded   
1612174                       0.0                 Fatigued/Drowsy   
1612175                       0.0                 Fatigued/Drowsy   
1612176                       0.0  Driver Inattention/Distraction   
1612177                       0.0                     Unspecified   

          CONTRIBUTING FACTOR VEHICLE 2 CONTRIBUTING FACTOR VEHICLE 3   
0                           Unspecified                           NaN  \
1                           Unspecified                           NaN   
2                                   NaN                           NaN   
3                           Unspecified                           NaN   
4        Driver Inattention/Distraction                           NaN   
...                                 ...                           ...   
1612173                     Unspecified                           NaN   
1612174                 Fatigued/Drowsy                   Unspecified   
1612175                     Unspecified                           NaN   
1612176                     Unspecified                   Unspecified   
1612177                             NaN                           NaN   

        CONTRIBUTING FACTOR VEHICLE 4 CONTRIBUTING FACTOR VEHICLE 5   
0                                 NaN                           NaN  \
1                                 NaN                           NaN   
2                                 NaN                           NaN   
3                                 NaN                           NaN   
4                                 NaN                           NaN   
...                               ...                           ...   
1612173                           NaN                           NaN   
1612174                           NaN                           NaN   
1612175                           NaN                           NaN   
1612176                           NaN                           NaN   
1612177                           NaN                           NaN   

                                      CONTRIBUTING FACTORS  
0        [Passing Too Closely, Unspecified, nan, nan, nan]  
1        [Passing Too Closely, Unspecified, nan, nan, nan]  
2        [Driver Inattention/Distraction, nan, nan, nan...  
3                [Unspecified, Unspecified, nan, nan, nan]  
4        [Driver Inattention/Distraction, Driver Inatte...  
...                                                    ...  
1612173  [Traffic Control Disregarded, Unspecified, nan...  
1612174  [Fatigued/Drowsy, Fatigued/Drowsy, Unspecified...  
1612175      [Fatigued/Drowsy, Unspecified, nan, nan, nan]  
1612176  [Driver Inattention/Distraction, Unspecified, ...  
1612177                  [Unspecified, nan, nan, nan, nan]  

[1612145 rows x 7 columns]

##### Dodanie kolumny z czynnikiem wypadków, wynikająca z połączenia wartości z 5 kolumn z czynnikami wypadków dla pojazdów 1,2,3,4,5

## Wyznaczenie kolumn potrzebnych do dalszej analizy

In [5]:
df = nypd_factors[['NUMBER OF PERSONS KILLED', "CONTRIBUTING FACTORS"]]
df

NUMBER OF PERSONS KILLED   
0                             0.0  \
1                             0.0   
2                             0.0   
3                             0.0   
4                             0.0   
...                           ...   
1612173                       0.0   
1612174                       0.0   
1612175                       0.0   
1612176                       0.0   
1612177                       0.0   

                                      CONTRIBUTING FACTORS  
0        [Passing Too Closely, Unspecified, nan, nan, nan]  
1        [Passing Too Closely, Unspecified, nan, nan, nan]  
2        [Driver Inattention/Distraction, nan, nan, nan...  
3                [Unspecified, Unspecified, nan, nan, nan]  
4        [Driver Inattention/Distraction, Driver Inatte...  
...                                                    ...  
1612173  [Traffic Control Disregarded, Unspecified, nan...  
1612174  [Fatigued/Drowsy, Fatigued/Drowsy, Unspecified...  
1612175      [Fatigued/Drowsy, Unspecified, nan, nan, nan]  
1612176  [Driver Inattention/Distraction, Unspecified, ...  
1612177                  [Unspecified, nan, nan, nan, nan]  

[1612145 rows x 2 columns]

##### Została utworzona tabela z kolumną osób zabitych oraz czynnikami wypadków 

## Zamiana wartości 0 w kolumnie z osobami zabitymi na wartość NaN, a następnie usunięcie tych wartości, gdyż nie będą one potrzebne do dalszej analizy

In [6]:
df['NUMBER OF PERSONS KILLED'] = df['NUMBER OF PERSONS KILLED'].replace(0, np.nan)
df

/tmp/ipykernel_87204/347875460.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['NUMBER OF PERSONS KILLED'] = df['NUMBER OF PERSONS KILLED'].replace(0, np.nan)


NUMBER OF PERSONS KILLED   
0                             NaN  \
1                             NaN   
2                             NaN   
3                             NaN   
4                             NaN   
...                           ...   
1612173                       NaN   
1612174                       NaN   
1612175                       NaN   
1612176                       NaN   
1612177                       NaN   

                                      CONTRIBUTING FACTORS  
0        [Passing Too Closely, Unspecified, nan, nan, nan]  
1        [Passing Too Closely, Unspecified, nan, nan, nan]  
2        [Driver Inattention/Distraction, nan, nan, nan...  
3                [Unspecified, Unspecified, nan, nan, nan]  
4        [Driver Inattention/Distraction, Driver Inatte...  
...                                                    ...  
1612173  [Traffic Control Disregarded, Unspecified, nan...  
1612174  [Fatigued/Drowsy, Fatigued/Drowsy, Unspecified...  
1612175      [Fatigued/Drowsy, Unspecified, nan, nan, nan]  
1612176  [Driver Inattention/Distraction, Unspecified, ...  
1612177                  [Unspecified, nan, nan, nan, nan]  

[1612145 rows x 2 columns]

In [7]:
df = df.dropna(subset=['NUMBER OF PERSONS KILLED'])
df

NUMBER OF PERSONS KILLED   
7                             1.0  \
467                           1.0   
540                           1.0   
708                           1.0   
911                           1.0   
...                           ...   
1608873                       1.0   
1609027                       1.0   
1610433                       1.0   
1610994                       1.0   
1611807                       1.0   

                                      CONTRIBUTING FACTORS  
7               [Unsafe Speed, Unspecified, nan, nan, nan]  
467      [Failure to Yield Right-of-Way, nan, nan, nan,...  
540      [Reaction to Uninvolved Vehicle, Unspecified, ...  
708      [Unsafe Speed, Unspecified, Unspecified, nan, ...  
911      [Driver Inattention/Distraction, nan, nan, nan...  
...                                                    ...  
1608873  [Aggressive Driving/Road Rage, Unspecified, na...  
1609027                  [Unspecified, nan, nan, nan, nan]  
1610433  [View Obstructed/Limited, Unspecified, nan, na...  
1610994  [Alcohol Involvement, Unspecified, nan, nan, nan]  
1611807          [Unspecified, Unspecified, nan, nan, nan]  

[1837 rows x 2 columns]

##### Oczyszczona tabela danych z conajmniej jedną ofiarą wypadków w NY

## Przypisanie ofiary do każdego z czynników wypadków

In [8]:
df_exploded = df.explode('CONTRIBUTING FACTORS')
df_exploded

NUMBER OF PERSONS KILLED CONTRIBUTING FACTORS
7                             1.0         Unsafe Speed
7                             1.0          Unspecified
7                             1.0                  NaN
7                             1.0                  NaN
7                             1.0                  NaN
...                           ...                  ...
1611807                       1.0          Unspecified
1611807                       1.0          Unspecified
1611807                       1.0                  NaN
1611807                       1.0                  NaN
1611807                       1.0                  NaN

[9185 rows x 2 columns]

##### Za pomocą metody explode, każdy czynnik wypadku został przypisany do ofiary

## Usunięcie wartości NaN z czynników wypadków

In [9]:
df_exploded_drop = df_exploded.dropna()
df_exploded_drop

NUMBER OF PERSONS KILLED            CONTRIBUTING FACTORS
7                             1.0                    Unsafe Speed
7                             1.0                     Unspecified
467                           1.0   Failure to Yield Right-of-Way
540                           1.0  Reaction to Uninvolved Vehicle
540                           1.0                     Unspecified
...                           ...                             ...
1610433                       1.0                     Unspecified
1610994                       1.0             Alcohol Involvement
1610994                       1.0                     Unspecified
1611807                       1.0                     Unspecified
1611807                       1.0                     Unspecified

[2893 rows x 2 columns]

##### Oczyszczona tabela danych z pustych wartości dla czynników wypadków

## Zsumowanie liczby osób zabitych do poszczególnego czynnika wypadku

In [10]:
person_of_killed = df_exploded_drop.groupby(['CONTRIBUTING FACTORS'])['NUMBER OF PERSONS KILLED'].sum()
person_of_killed

CONTRIBUTING FACTORS
Aggressive Driving/Road Rage                               11.0
Alcohol Involvement                                        61.0
Backing Unsafely                                           28.0
Driver Inattention/Distraction                            268.0
Driver Inexperience                                        24.0
Driverless/Runaway Vehicle                                  1.0
Drugs (Illegal)                                             2.0
Drugs (illegal)                                             4.0
Failure to Keep Right                                       3.0
Failure to Yield Right-of-Way                             189.0
Fatigued/Drowsy                                             3.0
Fell Asleep                                                13.0
Following Too Closely                                      24.0
Glare                                                       1.0
Illnes                                                     33.0
Lost Consciousness 

##### Tabela zawierającą zsumowaną liczbę ofiar dla każdego z czynnika wypadku

## Wyznaczenie 5 czynników, w których najcześciej występują ofiary śmiertelne wypadków w NY

In [17]:
top_data_of_killed = person_of_killed.groupby('CONTRIBUTING FACTORS').nlargest(5).reset_index(level=0, drop=True)
top_data_of_killed = top_data_of_killed.reset_index()

top_data_of_killed.sort_values('NUMBER OF PERSONS KILLED',ascending=False).head(5)


CONTRIBUTING FACTORS  NUMBER OF PERSONS KILLED
37                     Unspecified                    1750.0
3   Driver Inattention/Distraction                     268.0
9    Failure to Yield Right-of-Way                     189.0
33     Traffic Control Disregarded                     182.0
36                    Unsafe Speed                     156.0

##### Tabela przedstawia 5 najczęstsze przyczyny wypadków, w których jest najwięcej ofiar śmiertelnych

## Podsumowanie

## Po przeanalizowaniu danych wejściowych - najwięcej osób śmiertelnych występuje w wypadkach spowodowanymi:
### 1. Nieokreślone - 1750 ofiar
### 2. Nieuwaga/rozproszenie kierowcy - 268 ofiar
### 3. Nieustąpienie pierwszeństwa przejazdu - 189 ofiar
### 4. Przejazd na czerwonym świetle - 182 ofiary
### 5. Nadmierna prędkość - 156 ofiar